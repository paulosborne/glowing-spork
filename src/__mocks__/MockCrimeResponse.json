[
  {
    "category": "vehicle-crime",
    "location_type": "Force",
    "location": {
      "latitude": "52.590250",
      "street": {
        "id": 1269279,
        "name": "On or near Sports/recreation Area"
      },
      "longitude": "-2.130356"
    },
    "context": "",
    "outcome_status": {
      "category": "Investigation complete; no suspect identified",
      "date": "2019-03"
    },
    "persistent_id": "a76ce31fbbdda06c72f1ebc7b583eefc092c8498b097bfc879929afd6b3fc6c4",
    "id": 72264228,
    "location_subtype": "",
    "month": "2019-02"
  },
  {
    "category": "burglary",
    "location_type": "Force",
    "location": {
      "latitude": "52.590250",
      "street": {
        "id": 1269279,
        "name": "On or near Sports/recreation Area"
      },
      "longitude": "-2.130356"
    },
    "context": "",
    "outcome_status": {
      "category": "Investigation complete; no suspect identified",
      "date": "2019-03"
    },
    "persistent_id": "a76ce31fbbdda06c72f1ebc7b583eefc092c8498b097bfc879929afd6b3fc6c4",
    "id": 72264229,
    "location_subtype": "",
    "month": "2019-02"
  }
]
