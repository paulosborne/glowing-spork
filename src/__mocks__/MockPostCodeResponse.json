{
  "status": 200,
  "result": {
    "postcode": "WV1 4QR",
    "quality": 1,
    "eastings": 391207,
    "northings": 299163,
    "country": "England",
    "nhs_ha": "West Midlands",
    "longitude": -2.131226,
    "latitude": 52.590195,
    "european_electoral_region": "West Midlands",
    "primary_care_trust": "Wolverhampton City",
    "region": "West Midlands",
    "lsoa": "Wolverhampton 020F",
    "msoa": "Wolverhampton 020",
    "incode": "4QR",
    "outcode": "WV1",
    "parliamentary_constituency": "Wolverhampton South West",
    "admin_district": "Wolverhampton",
    "parish": "Wolverhampton, unparished area",
    "admin_county": null,
    "admin_ward": "St Peter's",
    "ced": null,
    "ccg": "NHS Wolverhampton",
    "nuts": "Wolverhampton",
    "codes": {
      "admin_district": "E08000031",
      "admin_county": "E99999999",
      "admin_ward": "E05001335",
      "parish": "E43000185",
      "parliamentary_constituency": "E14001051",
      "ccg": "E38000210",
      "ccg_id": "06A",
      "ced": "E99999999",
      "nuts": "UKG39"
    }
  }
}
