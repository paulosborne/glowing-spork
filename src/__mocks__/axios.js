export const mockPost = jest.fn();
export const mockGet = jest
  .fn()
  .mockImplementation(() => Promise.resolve({ data: undefined }));
export const mockRequest = jest.fn();
export const mockPut = jest.fn();
export const mockRequestInterceptor = jest.fn();

export const mockCreate = jest.fn().mockImplementation(() => ({
  get: mockGet,
  post: mockPost,
  put: mockPut,
  interceptors: {
    request: {
      use: mockRequestInterceptor,
    },
  },
}));

module.exports = {
  mockRequestInterceptor,
  create: mockCreate,
  post: mockPost,
  put: mockPut,
  get: mockGet,
};
