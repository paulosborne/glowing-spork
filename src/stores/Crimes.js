import { flow, types } from 'mobx-state-tree';
import Crime from './entities/Crime';
import CrimeCategory from './entities/CrimeCategory';
import PoliceData from '../services/PoliceData';
import PostCodeIO from '../services/PostCodeIO';
import Coordinates from './entities/Coordinates';

const Crimes = types
  .model('Crimes', {
    data: types.optional(types.array(Crime), []),
    categories: types.optional(types.map(CrimeCategory), {}),
    postcode: types.maybeNull(types.string),
    coordinates: types.optional(Coordinates, {}),
  })
  .views(self => ({
    get categoryList() {
      return Array.from(self.categories.values());
    },
    get selectedCategories() {
      return Array.from(self.categories.values())
        .filter(({ selected }) => selected)
        .map(({ url }) => url);
    },
    get visible() {
      return self.data.filter(crime =>
        self.selectedCategories.includes(crime.category),
      );
    },
  }))
  .actions(self => ({
    /**
     * Sets the coordinates
     * @param {object} params
     * @param {number} params.longitude
     * @param {number} params.latitude
     */
    setCoordinates({ longitude, latitude } = {}) {
      self.coordinates.setLatitude(latitude);
      self.coordinates.setLongitude(longitude);
    },

    /**
     * Gets additional information about a postcode
     * @param {string} postcode
     */
    getCoordinates: flow(function* (postcode) {
      const { data } = yield PostCodeIO.get(`/postcodes/${postcode}`);
      self.setCoordinates(data.result);
    }),

    /**
     * Fetches all crimes at the given coordinates for the given date.
     * @param {string} month
     * @param {number} longitude
     * @param {number} latitude
     */
    getCrimesAtLocation: flow(function* (year, month) {
      if (!self.coordinates.isValid) {
        return;
      }
      try {
        const { data } = yield PoliceData.get('/crimes-at-location', {
          params: {
            date: `${year}-${month}`,
            lng: self.coordinates.longitude,
            lat: self.coordinates.latitude,
          },
        });
        self.data.replace(data);
      } catch (err) {
        console.log({ err });
      }
    }),

    /**
     * Fetches the crime categories for the given year & month
     * @param {number} year
     * @param {number} month
     */
    getCrimeCategories: flow(function* (year, month) {
      const { data } = yield PoliceData.get(
        `/crime-categories?date=${year}-${month}`,
      );

      self.categories.clear();
      data.forEach(datum => self.categories.set(datum.url, datum));
    }),

    /**
     * Returns the category matching the given url
     * @param {string} url
     */
    getCategory(url) {
      return self.categories.get(url);
    },

    /**
     * Sets the current location for crime
     * @param {string} postcode
     */
    setCrimeLocation: flow(function* (postcode) {
      yield self.getCoordinates(postcode);
    }),
  }));

export default Crimes;
