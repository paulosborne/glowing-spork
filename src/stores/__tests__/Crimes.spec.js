import { isStateTreeNode, getSnapshot } from 'mobx-state-tree';
import Crimes from '../Crimes';
import axios from 'axios';
import MockCrimeCategoryResponse from 'MockCrimeCategoryResponse';
import MockCrimeResponse from 'MockCrimeResponse';
import MockPostCodeResponse from 'MockPostCodeResponse';

describe('Store: Crimes', () => {
  let instance;

  beforeEach(() => {
    instance = Crimes.create();
    jest.resetAllMocks();
  });

  it('should create the store correctly', () => {
    expect(isStateTreeNode(instance)).toEqual(true);
    expect(getSnapshot(instance)).toMatchSnapshot();
  });

  /**
   * getCrimeCategories()
   */
  describe('getCrimeCategories()', () => {
    it('should be defined', () => {
      expect(instance.getCrimeCategories).toBeDefined();
    });

    it('should fetch crime categories correctly', async () => {
      const month = 10;
      const year = 2020;
      axios.get.mockResolvedValue({ data: MockCrimeCategoryResponse });
      await instance.getCrimeCategories(year, month);
      expect(axios.get).toHaveBeenCalledWith(
        `/crime-categories?date=${year}-${month}`,
      );
      expect(instance.categories.size).toEqual(
        MockCrimeCategoryResponse.length,
      );
    });
  });

  /**
   * getCoordinates()
   */
  describe('getCoordinates()', () => {
    it('should be defined', () => {
      expect(instance.getCoordinates).toBeDefined();
    });

    it('should lookup the coordinates from the postcode correctly', async () => {
      const { postcode } = MockPostCodeResponse.result;
      axios.get.mockResolvedValue({ data: MockPostCodeResponse });
      instance.setCrimeLocation(postcode);
      await instance.getCoordinates();
      expect(axios.get).toHaveBeenCalledWith(`/postcodes/${postcode}`);
    });
  });

  /**
   * getCrimesAtLocation()
   */
  describe('getCrimesAtLocation()', () => {
    it('should be defined', () => {
      expect(instance.getCrimesAtLocation).toBeDefined();
    });

    it('should call the crimes-by-location API correctly', async () => {
      const year = 2020;
      const month = 10;
      const longitude = -1;
      const latitude = 53;

      axios.get.mockResolvedValue({ data: MockCrimeResponse });
      instance.setCoordinates({ longitude, latitude });
      await instance.getCrimesAtLocation(year, month);
      expect(axios.get).toHaveBeenCalledWith('/crimes-at-location', {
        params: {
          date: `${year}-${month}`,
          lng: longitude,
          lat: latitude,
        },
      });
      expect(getSnapshot(instance.data)).toMatchSnapshot();
    });
  });
});
