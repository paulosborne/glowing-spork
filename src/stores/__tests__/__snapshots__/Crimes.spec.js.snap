// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store: Crimes getCrimesAtLocation() should call the crimes-by-location API correctly 1`] = `
Array [
  Object {
    "category": "vehicle-crime",
    "context": "",
    "id": 72264228,
    "location": Object {
      "latitude": "52.590250",
      "longitude": "-2.130356",
      "street": Object {
        "name": "On or near Sports/recreation Area",
      },
    },
    "location_type": "Force",
    "month": "2019-02",
    "outcome_status": Object {
      "category": "Investigation complete; no suspect identified",
      "date": "2019-03",
    },
  },
  Object {
    "category": "burglary",
    "context": "",
    "id": 72264229,
    "location": Object {
      "latitude": "52.590250",
      "longitude": "-2.130356",
      "street": Object {
        "name": "On or near Sports/recreation Area",
      },
    },
    "location_type": "Force",
    "month": "2019-02",
    "outcome_status": Object {
      "category": "Investigation complete; no suspect identified",
      "date": "2019-03",
    },
  },
]
`;

exports[`Store: Crimes should create the store correctly 1`] = `
Object {
  "categories": Object {},
  "coordinates": Object {
    "latitude": null,
    "longitude": null,
  },
  "data": Array [],
  "postcode": null,
}
`;
